base/iterators.jl

```diff
@@ -739,11 +739,32 @@ julia> collect(Iterators.flatten((1:2, 8:9)))
 flatten(itr) = Flatten(itr)
 
 eltype(::Type{Flatten{I}}) where {I} = eltype(eltype(I))
-iteratorsize(::Type{Flatten{I}}) where {I} = SizeUnknown()
 iteratoreltype(::Type{Flatten{I}}) where {I} = _flatteneltype(I, iteratoreltype(I))
 _flatteneltype(I, ::HasEltype) = iteratoreltype(eltype(I))
 _flatteneltype(I, et) = EltypeUnknown()
 
+flatten_iteratorsize(::Union{HasShape, HasLength}, b::Type{<:Tuple}) = HasLength()
+flatten_iteratorsize(::Union{HasShape, HasLength}, b::Type{<:Number}) = HasLength()
+flatten_iteratorsize(a, b) = SizeUnknown()
+
+function flatten_length(f, ::Type{<:Tuple})
+    if !isleaftype(T)
+        throw(ArgumentError(
+            "Cannot compute length of a tuple-type which is not a leaf-type"))
+    end
+    nfields(T)*length(f.it)
+end
+flatten_length(f, ::Type{T}) where {T<:Number} = length(f.it)
+length(f::Flatten{I}) where {I} = flatten_length(f, eltype(I))
```

Comment:
For non-(number, tuple), this will throw a `MethodError` right? why not a fall-ball `flatten_length` which throws a more informative message?